### PR TITLE
implement an user-friendly API

### PR DESCRIPTION
**Pull request by andylokandy**
Refactored API and some of the internal stuff.

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
 
 #[inline]
-pub(crate) fn real_time_ns() -> u64 {
+pub fn real_time_ns() -> u64 {
```
Comment (zhongzc) [inline src/utils.rs:4]: Why `pub`?

Comment (andylokandy) [reply src/utils.rs:4]: It will not be exported, because the mod it‘s in is private.

File: src/lib.rs [hunk 2]
```diff
@@ -153,3 +146,47 @@ pub struct Properties {
     pub property_lens: Vec<u64>,
     pub payload: Vec<u8>,
 }
+
+pub fn start_trace<T: Into<u32>>(event: T) -> Option<(ScopeGuard, Collector)> {
```
Comment (zhongzc) [inline src/lib.rs:150]: I can image the usage pattern:
```rust
let (guard, collector) = if ENABLE_TRACE {
  if let Some((guard, collector)) = start_trace(EVENT) {
    (Some(guard), Some(collector))
  } else {
    (None, None)
  }
} else {
  (None, None)
};
```

**Comment by zhongzc** [comment]
`README.md` needs to change also.

File: src/lib.rs (hunk 2 above)
Comment (andylokandy) [reply src/lib.rs]: done

**Comment by zhongzc** [comment]
Please fix build

```
cargo build --workspace --all-targets
```

File: src/lib.rs [hunk 3]
```diff
@@ -1,37 +1,30 @@
...
+mod collector;
+mod trace;
+mod utils;
 
-pub use collector::*;
-pub use trace::*;
-pub use trace_async::*;
-pub use trace_local::*;
+pub use collector::Collector;
+pub use trace::{ScopeGuard, SpanGuard};
```
Comment (zhongzc) [inline src/lib.rs:13]: `pub use thread`?

File: src/thread.rs [hunk 4]
```diff
@@ -44,45 +63,54 @@ impl TraceHandle {
...
                     elapsed_cycles: now_cycles.saturating_sub(inner.suspending_begin_cycles),
-                    event: pending_event,
+                    event,
                 },
-                settle_event,
+                event,
             ) {
                 // Reserve these for the next suspending process
-                inner.next_suspending_state = crate::State::Scheduling;
                 inner.next_related_id = self_id;
```
Comment (zhongzc) [inline src/thread.rs:109]:
```diff
+                inner.next_suspending_state = State::Scheduling;
```

Comment (andylokandy) [reply src/thread.rs:110]: done

File: src/thread.rs [hunk 5]
```diff
@@ -1,41 +1,60 @@
...
-struct AsyncTraceInner {
-    collector: std::sync::Arc<crate::collector::CollectorInner>,
-    next_suspending_state: crate::State,
+use crate::collector::CollectorInner;
+use crate::{ScopeGuard, State};
+
+/// Bind the current tracing context to another executing context (e.g. a closure).
+///
+/// ```no_run
+/// # use minitrace::new_async_scope;
```
Comment (zhongzc) [inline src/thread.rs:9]: /// ```
/// # use minitrace::thread::new_async_scope;

**Review by zhongzc — APPROVED** [review]

**Pull request closed without merging**